client/network/common/src/config.rs

```diff
@@ -218,6 +243,8 @@ pub struct NonDefaultSetConfig {
 	/// If a fallback is used, it will be reported in
 	/// `sc_network::protocol::event::Event::NotificationStreamOpened::negotiated_fallback`
 	pub fallback_names: Vec<protocol::ProtocolName>,
+	/// Handshake of the protocol
+	pub handshake: Option<NotificationHandshake>,
```

Comment:
For what it's worth I'm really not sure about this field. Looking at the code, None means "default", which means "the node's role".
Intuitively, to me, when I see a field named `handshake` which can take an `Option`, I assume that `None` means something like "no handshake".
It is also worth mentioning that while the handshake being sent is now customizable, the handshake decoding is still hardcoded. This is the reason why the handshake was so far not customizable.

At the very least, this field should IMO be very well documented, just like all the other fields of the config as well documented.

Reply:
This PR was not about addressing the customizable handshakes and the implementation for `NotificationHandshake` as right now stands is not the final version of it. I have been looking into how to make it totally generic but I don't have anything concrete to show yet. I plan to work on that when the block announcement protocol is converted to a generic notification protocol and the syncing code for the most part is out of `Protocol` but there is still some stuff I need to do before I can start working on that. This PR was about getting the block announcement protocol out of `Protocol` so that `ChainSync` can be the owner of that protocol.

But yes, it's not perfect a solution and I don't mind changing the type temporarily to an enum though, if it makes things more clear.